storage: fix (the semantics of) MVCCStats.GCBytesAge

tbg opened the pull request:
The semantics for computing GCBytesAge were incorrect and are fixed in
this commit. Prior to this commit, a non-live write would accrue
GCBytesAge from its own timestamp on. That is, if you wrote two versions
of a key at 1s and 2s, then when the older version is replaced (at 2s)
it would start out with one second of age (from 1s to 2s). However, the
key *really* became non-live at 2s, and should have had an age of zero.

By extension, imagine a table with lots of writes all dating back to
early 2017, and assume that today (early 2018) all these writes are
deleted (i.e. a tombstone placed on top of them). Prior to this commit,
each key would immediately get assigned an age of `(early 2018) - early
2017)`, i.e. a very large number. Yet, the GC queue could only garbage
collect them after (early 2018) + TTL`, so by default 25 hours after
the deletion. We use GCBytesAge to trigger the GC queue, so that would
cause the GC queue to run without ever getting to remove anything, for
the TTL. This was a big problem bound to be noticed by users.

This commit changes the semantics to what the GCQueue (and the layman)
expects:

1. when a version is shadowed, it becomes non-live at that point and
   also starts accruing GCBytesAge from that point on.
2. deletion tombstones are an exception: They accrue age from their
   own timestamp on. This makes sense because a tombstone can be
   deleted whenever it's older than the TTL (as opposed to a value,
   which can only be deleted once it's been *shadowed* for longer
   than the TTL).

This work started out by updating `ComputeStatsGo` to have the desired
semantics, fixing up existing tests, and then stress testing
`TestMVCCStatsRandomized` with short history lengths to discover failure
modes which were then transcribed into small unit tests. When no more
such failures were discoverable, the resulting logic in the various
incremental MVCCStats update helpers was simplified and documented, and
`ComputeStats` updated accordingly. In turn, `TestMVCCStatsBasic` was
removed: it was notoriously hard to read and maintain, and does not add
any coverage at this point.

The recomputation of the stats in existing clusters is addressed in
https://github.com/cockroachdb/cockroach/pull/21345.

Fixes https://github.com/cockroachdb/cockroach/issues/20554.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21407)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/engine/mvcc.go, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2v7-622HqARsrc0zyt:-L2v7-622HqARsrc0zyu:b-8gkfnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L234)):*
> ```Go
> 	var ms enginepb.MVCCStats
> 
> 	if false {
> ```

Remove this before merging?

---

*[pkg/storage/engine/mvcc.go, line 255 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2v7n9pXNLRZeeTOC21:-L2v7n9pXNLRZeeTOC22:b-6itqtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L255)):*
> ```Go
> 
> 	if isSysLocal(key) {
> 		// Handling system-local keys is straightforward because no ageable quantities
> ```

If there are transactions and intents (which we handle below), there are age-able quantities. Specifically, the range-local copy of the range descriptor has historical revisions that need to be GC'd. Maybe it's small enough to ignore that for the purposes of GCBytesAge, but the comment should indicate that that's what it's doing. 

---

*[pkg/storage/engine/mvcc.go, line 413 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2vFPOynuWoD0O5GWqE:-L2vFPOynuWoD0O5GWqF:b-wwbdlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L413)):*
> ```Go
> 	var ms enginepb.MVCCStats
> 
> 	if false {
> ```

Either delete this or put in a comment about why this disabled code is here. 

---

*[pkg/storage/engine/mvcc.go, line 446 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2vF_m7o91n4qzA8DXq:-L2vF_m7o91n4qzA8DXr:b-hrxshl) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L446)):*
> ```Go
> 	// Jump into the method below for extensive commentary on their semantics
> 	// and "rules one and two".
> 	_ = updateStatsOnPut
> ```

Nice hyperlink :)

---

*[pkg/storage/engine/mvcc.go, line 2666 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2v61VZClusnrkujCXm:-L2v61V_UCfKIVqDPaux:b-im5otl) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L2666)):*
> ```Go
> 	var prevKey []byte
> 	first := false
> 	var deletionNanos int64
> ```

s/deletionNanos/overwriteNanos/?

This could also use a comment about how we're iterating in reverse chronological order, so this tracks the timestamp of a value for use in computing the GC timestamp of its predecessor.

---

*[pkg/storage/engine/mvcc.go, line 2702 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2vGfv8V1ErW1w0AF_u:-L2vGfv9-32b8holaJZU:bkcev4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L2702)):*
> ```Go
> 
> 		if !isValue || implicitMeta {
> 			metaKeySize := int64(len(unsafeKey.Key)) + 1 // unsafeKey.EncodedSize()
> ```

What's up with this comment? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-:-L2vH-GN0WCEtlBZj6SA:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 3 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/engine/mvcc.go, line 227 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L3-8IqKtkovGVfcBgTm:-L3-8IqKtkovGVfcBgTn:b497nhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L227)):*
> ```Go
> func updateStatsOnPut(
> 	key roachpb.Key,
> 	prevKeySize, prevValSize int64,
> ```

It doesn't look like we need `prevKeySize`. In fact, it only looks like we need `prevValSize` to determine whether the previous value is a deletion tombstone. Should we trim these arguments? Perhaps just down to `prevIsTombstone bool`? That would also make a lot of the `if prevValSize > 0 {` checks easier to read.

---

*[pkg/storage/engine/mvcc.go, line 405 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L3-8_-Aik_1HnUcFpt8:-L3-8_-Aik_1HnUcFpt9:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L405)):*
> ```Go
> func updateStatsOnResolve(
> 	key roachpb.Key,
> 	prevKeySize, prevValSize int64,
> ```

Ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/storage/engine/mvcc.go, line 227 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L3-8IqKtkovGVfcBgTm:-L3-LAgKJWmhq7lQuaea:b5d5q20) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L227)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It doesn't look like we need `prevKeySize`. In fact, it only looks like we need `prevValSize` to determine whether the previous value is a deletion tombstone. Should we trim these arguments? Perhaps just down to `prevIsTombstone bool`? That would also make a lot of the `if prevValSize > 0 {` checks easier to read.
</blockquote></details>

We use prevValSize for arithmetic so I wasn't able to turn it into a bool, but we were not using `prevKeySize` and I removed it. I'm also using a local bool to simplify reading the logic, see if that makes more sense now.

---

*[pkg/storage/engine/mvcc.go, line 255 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2v7n9pXNLRZeeTOC21:-L3-IkwVAKqk7uUUJxCA:b-5zlvlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L255)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If there are transactions and intents (which we handle below), there are age-able quantities. Specifically, the range-local copy of the range descriptor has historical revisions that need to be GC'd. Maybe it's small enough to ignore that for the purposes of GCBytesAge, but the comment should indicate that that's what it's doing. 
</blockquote></details>

I clarified that we just don't track them but could.

---

*[pkg/storage/engine/mvcc.go, line 405 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L3-8_-Aik_1HnUcFpt8:-L3-LK8fSTKP3aYqUbY0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L405)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.

---

*[pkg/storage/engine/mvcc.go, line 413 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2vFPOynuWoD0O5GWqE:-L3-ILa8GAhWNBimnJ1N:bj4fp77) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L413)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Either delete this or put in a comment about why this disabled code is here. 
</blockquote></details>

Done, :crossed_fingers: that we won't need it any more.

---

*[pkg/storage/engine/mvcc.go, line 2666 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2v61VZClusnrkujCXm:-L3-Hbw7hExkWxIiMwRo:b-kgqao7) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L2666)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/deletionNanos/overwriteNanos/?

This could also use a comment about how we're iterating in reverse chronological order, so this tracks the timestamp of a value for use in computing the GC timestamp of its predecessor.
</blockquote></details>

Added comment. Renamed to `accrueGCAgeNanos` which I like because it specifies what we're using it for.

---

*[pkg/storage/engine/mvcc.go, line 2702 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2vGfv8V1ErW1w0AF_u:-L3-GM2xrNiK2rA0uLm7:buojxv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L2702)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What's up with this comment? 
</blockquote></details>

I toyed around with using `EncodedSize()` here instead, but decided it wasn't worth it. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/engine/mvcc.go, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L2v7-622HqARsrc0zyt:-L3-Mn5tBv7qml9lKtyl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/63cd5c4e34914921b3ef24b7ad4704a2957b4b12/pkg/storage/engine/mvcc.go#L234)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Remove this before merging?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[c-deps/libroach/db.cc, line 2361 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L3-efGHRDMxOPfymzBA:-L3-efGHRDMxOPfymzBB:be7jsin) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3977cbca585609172b94fd0ceb438433b4e753f/c-deps/libroach/db.cc#L2361)):*
> ```cpp
>   // NB: making this uninitialized triggers compiler warnings
>   // with `-Werror=maybe-uninitialized`. This warning seems like
>   // a false positive (changing the above line to `first=false`
> ```

The above line *is* `first=false`, so maybe this comment meant something else?

---

*[c-deps/libroach/db.cc, line 2456 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L3-ercRqsT3N5BUwqyW:-L3-ercRqsT3N5BUwqyX:b-fenjpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3977cbca585609172b94fd0ceb438433b4e753f/c-deps/libroach/db.cc#L2456)):*
> ```cpp
>           stats.gc_bytes_age += total_bytes * age_factor(wall_time, now_nanos);
>         } else {
>           stats.gc_bytes_age += total_bytes * age_factor(accrue_gc_age_nanos, now_nanos);
> ```

Maybe assert here that `accrue_gc_age_nanos` is non-zero to ensure we didn't hit the uninitialized path that the compiler was warning about. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: fix (the semantics of) MVCCStats.GCBytesAge

The semantics for computing GCBytesAge were incorrect and are fixed in
this commit. Prior to this commit, a non-live write would accrue
GCBytesAge from its own timestamp on. That is, if you wrote two versions
of a key at 1s and 2s, then when the older version is replaced (at 2s)
it would start out with one second of age (from 1s to 2s). However, the
key *really* became non-live at 2s, and should have had an age of zero.

By extension, imagine a table with lots of writes all dating back to
early 2017, and assume that today (early 2018) all these writes are
deleted (i.e. a tombstone placed on top of them). Prior to this commit,
each key would immediately get assigned an age of `(early 2018) - early
2017)`, i.e. a very large number. Yet, the GC queue could only garbage
collect them after (early 2018) + TTL`, so by default 25 hours after
the deletion. We use GCBytesAge to trigger the GC queue, so that would
cause the GC queue to run without ever getting to remove anything, for
the TTL. This was a big problem bound to be noticed by users.

This commit changes the semantics to what the GCQueue (and the layman)
expects:

1. when a version is shadowed, it becomes non-live at that point and
   also starts accruing GCBytesAge from that point on.
2. deletion tombstones are an exception: They accrue age from their
   own timestamp on. This makes sense because a tombstone can be
   deleted whenever it's older than the TTL (as opposed to a value,
   which can only be deleted once it's been *shadowed* for longer
   than the TTL).

This work started out by updating `ComputeStatsGo` to have the desired
semantics, fixing up existing tests, and then stress testing
`TestMVCCStatsRandomized` with short history lengths to discover failure
modes which were then transcribed into small unit tests. When no more
such failures were discoverable, the resulting logic in the various
incremental MVCCStats update helpers was simplified and documented, and
`ComputeStats` updated accordingly. In turn, `TestMVCCStatsBasic` was
removed: it was notoriously hard to read and maintain, and does not add
any coverage at this point.

The recomputation of the stats in existing clusters is addressed in
https://github.com/cockroachdb/cockroach/pull/21345.

Fixes https://github.com/cockroachdb/cockroach/issues/20554.

Release note (bug fix): fix a problem that could cause spurious GC
activity, in particular after dropping a table.

tschottdorf commented:
Review status: 3 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[c-deps/libroach/db.cc, line 2361 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L3-efGHRDMxOPfymzBA:-L3-fM-B0SZz_7al1aJi:bai9vaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3977cbca585609172b94fd0ceb438433b4e753f/c-deps/libroach/db.cc#L2361)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The above line *is* `first=false`, so maybe this comment meant something else?
</blockquote></details>

Nope, just wanted to write `true`. Fixed.

---

*[c-deps/libroach/db.cc, line 2456 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407#-L3-ercRqsT3N5BUwqyW:-L3-fg4kNzkXTrX5_dYN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3977cbca585609172b94fd0ceb438433b4e753f/c-deps/libroach/db.cc#L2456)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Maybe assert here that `accrue_gc_age_nanos` is non-zero to ensure we didn't hit the uninitialized path that the compiler was warning about. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21407)*
<!-- Sent from Reviewable.io -->
